docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,210 @@
+
+Train a Model
+#############
+
+.. code-block:: bash
+
+   rasa train
+
+Training a model verifies that your NLU pipeline and policy configurations are
+valid and trainable, and it provides a model to use for test conversations. 
+If your the passes the CI tests, then you can also upload the trained model 
+to your server as :ref:`part of the continuous deployment
+process <uploading-a-model>`. 
+
+Test the Assistant
+##################
+
+Testing your trained model on :ref:`test conversations
+<end-to-end-testing>` is the best way to have confidence in how your assistant
+will act in certain situations. These stories, written in a modified story
+format, allow you to provide entire conversations and test that, given this
+user input, your model will behave in the expected manner. This is especially
+important as you start introducing more complicated stories from user
+conversations. End-to-end testing is only as thorough and accurate as the test
+cases you include, so you should continue to grow your set of test conversations
+as you make improvements to your assistant. Your test conversations
+should be representative of the true distribution of real conversations. 
+Rasa X makes it easy to `add test conversations based on real conversations <https://rasa.com/docs/rasa-x/user-guide/improve-assistant.html#add-test-conversation>`_.
+
+.. code-block:: bash
+
+   rasa test --stories tests/conversation_tests.md --fail-on-prediction-errors
+
+The ``--fail-on-prediction-errors`` flag ensures the test will fail if any test conversation fails.
+
+Note: End-to-end testing does **not** execute your action code. You will need to
+:ref:`test your action code <testing-action-code>` in a seperate step.
+
+Compare NLU Performance
+#######################
+
+If you've made significant changes to your NLU training data (such as adding or
+splitting intents, or adding/changing a lot of examples), you should run a
+:ref:`full NLU evaluation <nlu-evaluation>`. You'll want to compare
+the performance of the NLU model without your changes to an NLU model with your
+changes. 
+
+You can do this by running NLU testing in cross-validation mode:
+
+.. code-block:: bash
+
+   rasa test nlu --cross-validation
+
+or by training a model on a training set and testing it on a test set. If you use the train-test
+set approach, it is best to :ref:`shuffle and split your data <train-test-split>` using ``rasa data split`` as part of this CI step, as
+opposed to using a static NLU test set, which can easily become outdated. 
+
+It's good practice to have this step produce an artifact that can be reviewed
+before changes are merged in. For example, `this worflow <https://gist.github.com/amn41/de555c93913a01fbd56df2e2d211862c>`_
+includes commenting on a PR with an NLU cross-validation results table.
+
+Since NLU comparison can be a fairly resource intensive test, you can set this
+test to run only when a certain tag (e.g. "NLU testing required") is present,
```

Comment:
Still not sure what tag means here

Reply:
sorry didn't see previous comment - here it means something like github  labels. How about `run only when a pull request is labelled as e.g. "NLU testing required"`

Reply:
Not everyone uses GitHub, so we should explain that in more detail and that is just an example of how you might explicitly tell those tests to run

Reply:
If this is explained in initial mention of tags (above), we could say here `run this test only when certain conditions, like changes to NLU data files or the presence of manual labels,  are met`